lib/tree.rb

```diff
@@ -1,64 +1,126 @@
 class TreeNode
   attr_reader :key, :value
   attr_accessor :left, :right
-
-   def initialize(key, val)
+  
+  def initialize(key, val)
     @key = key
     @value = val
     @left = nil
     @right = nil
-   end
+  end
 end
 
 class Tree
   attr_reader :root
   def initialize
     @root = nil
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  # Time Complexity: O(log n)
+  # Space Complexity: O(1)
+  def add_helper(current_node, key, value)
+    return TreeNode.new(key, value) if current_node.nil?
+    
+    if key < current_node.key
+      current_node.left = add_helper(current_node.left, key, value)
+    else
+      current_node.right = add_helper(current_node.right, key, value)
+    end 
+    
+    return current_node
+  end 
+  
   def add(key, value)
-    raise NotImplementedError
+    @root = add_helper(@root, key, value)
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  # Time Complexity: O(log n)
+  # Space Complexity: O(1)  
+  def find_helper(current_node, key)
+    return nil if current_node.nil?
+    
+    if key > current_node.key 
+      current_node.right = find_helper(current_node.right, key)
+    elsif key < current_node.key
+      current_node.left = find_helper(current_node.left, key)
+    else
+      return current_node.value
+    end
+  end
+  
   def find(key)
-    raise NotImplementedError
+    return find_helper(@root, key)
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)   
   def inorder
-    raise NotImplementedError
+    return inorder_helper(@root, [])
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def inorder_helper(current_node, list)
+    return list if current_node.nil? 
+    
+    inorder_helper(current_node.left, list)
+    list << {key: current_node.key, value: current_node.value}
+    
+    inorder_helper(current_node.right, list)
+    return list
+  end
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def preorder
-    raise NotImplementedError
+    return pre_order_helper(@root, [])
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def pre_order_helper(current_node, list)
+    return list if current_node.nil? 
+    
+    list << {key: current_node.key, value: current_node.value}
+    pre_order_helper(current_node.left, list)
+    pre_order_helper(current_node.right, list)
+    return list
+  end
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def postorder
-    raise NotImplementedError
+    return post_order_helper(@root, [])
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def post_order_helper(current_node, list)
+    return list if current_node.nil? 
+    
+    post_order_helper(current_node.left, list)
+    post_order_helper(current_node.right, list)
+    list << {key: current_node.key, value: current_node.value}
+    return list
+  end
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)  
+  def height_helper(current_node)
```

Comment:
The space complexity here depends on the recursive calls which depend on the height of the tree.  So the space complexity is O(log n) for balanced trees and O(n) for unbalanced ones. 

Otherwise the method is quite well done!  😃

Reply:
Fixed.